File: backend/src/routes/trades.js
const express = require('express');
const router = express.Router();
const { authenticate, optionalAuth } = require('../middleware/auth');
const { queryOne, queryAll, query } = require('../config/database');

/**
 * GET /api/journal/trades
 * Get all trades for the authenticated account
 */
router.get('/', authenticate, async (req, res) => {
    try {
        const { status, pair, startDate, endDate, limit = 100, offset = 0 } = req.query;

        let queryText = `
      SELECT * FROM trades 
      WHERE account_id = $1
    `;
        const params = [req.user.accountId];
        let paramCount = 1;

        // Add filters
        if (status) {
            paramCount++;
            queryText += ` AND status = $${paramCount}`;
            params.push(status);
        }

        if (pair) {
            paramCount++;
            queryText += ` AND pair = $${paramCount}`;
            params.push(pair);
        }

        if (startDate) {
            paramCount++;
            queryText += ` AND trade_date >= $${paramCount}`;
            params.push(startDate);
        }

        if (endDate) {
            paramCount++;
            queryText += ` AND trade_date <= $${paramCount}`;
            params.push(endDate);
        }

        queryText += ` ORDER BY open_time DESC LIMIT $${paramCount + 1} OFFSET $${paramCount + 2}`;
        params.push(limit, offset);

        const trades = await queryAll(queryText, params);
        res.json(trades);
    } catch (error) {
        console.error('❌ Get trades error:', error);
        res.status(500).json({ error: 'Failed to fetch trades' });
    }
});

/**
 * GET /api/journal/trades/:id
 * Get specific trade
 */
router.get('/:id', authenticate, async (req, res) => {
    try {
        const trade = await queryOne(
            'SELECT * FROM trades WHERE id = $1 AND account_id = $2',
            [req.params.id, req.user.accountId]
        );

        if (!trade) {
            return res.status(404).json({ error: 'Trade not found' });
        }

        res.json(trade);
    } catch (error) {
        console.error('❌ Get trade error:', error);
        res.status(500).json({ error: 'Failed to fetch trade' });
    }
});

/**
 * POST /api/journal/trades
 * Create new trade (manual or via MT5 EA webhook)
 * Supports accountNumber authentication for webhook
 */
router.post('/', async (req, res) => {
    try {
        const { accountNumber } = req.body;

        // Authenticate via token or accountNumber
        let userId, accountId;

        if (req.headers.authorization) {
            // Authenticated via JWT
            const { authenticate } = require('../middleware/auth');
            await new Promise((resolve, reject) => {
                authenticate(req, res, (err) => {
                    if (err) reject(err);
                    else resolve();
                });
            });
            userId = req.user.userId;
            accountId = req.user.accountId;
        } else if (accountNumber) {
            // Authenticate via accountNumber (EA webhook)
            const account = await queryOne(
                'SELECT id, user_id FROM mt5_accounts WHERE account_number = $1',
                [accountNumber]
            );

            if (!account) {
                return res.status(404).json({ error: 'MT5 account not found' });
            }

            userId = account.user_id;
            accountId = account.id;
        } else {
            return res.status(401).json({ error: 'Authentication required' });
        }

        const {
            event, // ORDER_OPEN or ORDER_CLOSE
            mt5Ticket,
            pair,
            direction,
            volume,
            entryPrice,
            exitPrice,
            stopLoss,
            takeProfit,
            profitLoss,
            commission,
            swap,
            openTime,
            closeTime,
            tradeDate,
            session,
            notes,
            tags,
            positionId,
        } = req.body;

        // Handle MT5 EA events
        if (event === 'ORDER_OPEN') {
            // Create or update trade as OPEN
            const existingTrade = mt5Ticket ? await queryOne(
                'SELECT id FROM trades WHERE mt5_ticket = $1 AND account_id = $2',
                [mt5Ticket, accountId]
            ) : null;

            if (existingTrade) {
                // Update existing trade
                const result = await queryOne(
                    `UPDATE trades SET
            pair = $1, direction = $2, volume = $3, entry_price = $4,
            stop_loss = $5, take_profit = $6, open_time = $7, 
            trade_date = $8, session = $9, status = 'OPEN', source = 'MT5'
           WHERE id = $10
           RETURNING *`,
                    [pair, direction, volume, entryPrice, stopLoss, takeProfit,
                        openTime, tradeDate, session, existingTrade.id]
                );
                return res.status(200).json(result);
            } else {
                // Create new OPEN trade
                const result = await queryOne(
                    `INSERT INTO trades (
            account_id, user_id, mt5_ticket, mt5_position_id, pair, direction, volume,
            entry_price, stop_loss, take_profit, open_time, trade_date, session,
            status, source
          ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, 'OPEN', 'MT5')
          RETURNING *`,
                    [accountId, userId, mt5Ticket, positionId, pair, direction, volume,
                        entryPrice, stopLoss, takeProfit, openTime, tradeDate, session]
                );
                return res.status(201).json(result);
            }
        } else if (event === 'ORDER_CLOSE') {
            // Find and update OPEN trade to CLOSED
            // Use positionId to find the trade (since close deal has different ticket)
            const existingTrade = positionId ? await queryOne(
                'SELECT id FROM trades WHERE mt5_position_id = $1 AND account_id = $2 AND status = $3',
                [positionId, accountId, 'OPEN']
            ) : null;

            if (existingTrade) {
                // Update to CLOSED
                const result = await queryOne(
                    `UPDATE trades SET
            exit_price = $1, close_time = $2, profit_loss = $3,
            commission = $4, swap = $5, status = 'CLOSED'
           WHERE id = $6
           RETURNING *`,
                    [exitPrice, closeTime, profitLoss, commission || 0, swap || 0, existingTrade.id]
                );
                return res.status(200).json(result);
            } else {
                // No matching OPEN trade found - skip this CLOSE event
                // Historical sync will handle it properly later
                console.log(`⚠️ Skipping ORDER_CLOSE for position ${positionId} - no matching OPEN trade found (will be synced via historical sync)`);
                return res.status(200).json({ 
                    message: 'CLOSE event skipped - no matching OPEN trade',
                    positionId: positionId,
                    note: 'Will be handled by historical sync'
                });
            }
        } else {
            // Manual trade creation
            const result = await queryOne(
                `INSERT INTO trades (
          account_id, user_id, pair, direction, volume, entry_price, exit_price,
          stop_loss, take_profit, open_time, close_time, trade_date, session,
          profit_loss, commission, swap, notes, tags, status, source
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, 'MANUAL')
        RETURNING *`,
                [accountId, userId, pair, direction, volume, entryPrice, exitPrice,
                    stopLoss, takeProfit, openTime, closeTime, tradeDate, session,
                    profitLoss, commission, swap, notes, tags, exitPrice ? 'CLOSED' : 'OPEN']
            );
            return res.status(201).json(result);
        }
    } catch (error) {
        console.error('❌ Create trade error:', error);
        res.status(500).json({ error: 'Failed to create trade' });
    }
});

/**
 * PUT /api/journal/trades/:id
 * Update existing trade
 */
router.put('/:id', authenticate, async (req, res) => {
    try {
        const {
            pair, direction, volume, entry_price, exit_price,
            stop_loss, take_profit, open_time, close_time, trade_date,
            session, profit_loss, commission, swap, notes, tags, status
        } = req.body;

        const result = await queryOne(
            `UPDATE trades SET
        pair = COALESCE($1, pair),
        direction = COALESCE($2, direction),
        volume = COALESCE($3, volume),
        entry_price = COALESCE($4, entry_price),
        exit_price = COALESCE($5, exit_price),
        stop_loss = COALESCE($6, stop_loss),
        take_profit = COALESCE($7, take_profit),
        open_time = COALESCE($8, open_time),
        close_time = COALESCE($9, close_time),
        trade_date = COALESCE($10, trade_date),
        session = COALESCE($11, session),
        profit_loss = COALESCE($12, profit_loss),
        commission = COALESCE($13, commission),
        swap = COALESCE($14, swap),
        notes = COALESCE($15, notes),
        tags = COALESCE($16, tags),
        status = COALESCE($17, status)
      WHERE id = $18 AND account_id = $19
      RETURNING *`,
            [pair, direction, volume, entry_price, exit_price, stop_loss, take_profit,
                open_time, close_time, trade_date, session, profit_loss, commission, swap,
                notes, tags, status, req.params.id, req.user.accountId]
        );

        if (!result) {
            return res.status(404).json({ error: 'Trade not found' });
        }

        res.json(result);
    } catch (error) {
        console.error('❌ Update trade error:', error);
        res.status(500).json({ error: 'Failed to update trade' });
    }
});

/**
 * DELETE /api/journal/trades/:id
 * Delete trade
 */
router.delete('/:id', authenticate, async (req, res) => {
    try {
        const result = await query(
            'DELETE FROM trades WHERE id = $1 AND account_id = $2',
            [req.params.id, req.user.accountId]
        );

        if (result.rowCount === 0) {
            return res.status(404).json({ error: 'Trade not found' });
        }

        res.json({ message: 'Trade deleted successfully' });
    } catch (error) {
        console.error('❌ Delete trade error:', error);
        res.status(500).json({ error: 'Failed to delete trade' });
    }
});

/**
 * DELETE /api/journal/trades
 * Bulk delete trades
 */
router.delete('/', authenticate, async (req, res) => {
    try {
        const { ids } = req.body;

        if (!ids || !Array.isArray(ids) || ids.length === 0) {
            return res.status(400).json({ error: 'Trade IDs required' });
        }

        const placeholders = ids.map((_, i) => `$${i + 2}`).join(',');
        const result = await query(
            `DELETE FROM trades WHERE account_id = $1 AND id IN (${placeholders})`,
            [req.user.accountId, ...ids]
        );

        res.json({
            message: 'Trades deleted successfully',
            deletedCount: result.rowCount
        });
    } catch (error) {
        console.error('❌ Bulk delete error:', error);
        res.status(500).json({ error: 'Failed to delete trades' });
    }
});

/**
 * POST /api/journal/trades/sync
 * Sync multiple trades from MT5 (for initial sync or recovery)
 * Supports accountNumber authentication
 */
router.post('/sync', async (req, res) => {
    try {
        const { accountNumber, trades } = req.body;

        if (!accountNumber) {
            return res.status(400).json({ error: 'Account number required' });
        }

        if (!trades || !Array.isArray(trades) || trades.length === 0) {
            return res.status(400).json({ error: 'Trades array required' });
        }

        // Find account by accountNumber
        const account = await queryOne(
            'SELECT id, user_id FROM mt5_accounts WHERE account_number = $1',
            [accountNumber]
        );

        if (!account) {
            return res.status(404).json({ error: 'Account not found' });
        }

        let syncedCount = 0;
        let skippedCount = 0;
        let errorCount = 0;

        for (const trade of trades) {
            try {
                const {
                    ticket, pair, direction, volume, entryPrice, exitPrice,
                    stopLoss, takeProfit, openTime, closeTime, tradeDate, session,
                    profitLoss, grossProfit, netProfit, commission, swap, notes, tags
                } = trade;

                if (!ticket) {
                    errorCount++;
                    continue;
                }

                // Check if trade already exists
                const existing = await queryOne(
                    'SELECT id FROM trades WHERE mt5_ticket = $1 AND account_id = $2',
                    [ticket, account.id]
                );

                // Use UPSERT to prevent duplicates
                // If mt5_ticket exists, update; otherwise insert
                await query(
                    `INSERT INTO trades (
                        account_id, user_id, mt5_ticket, mt5_position_id, pair, direction, volume,
                        entry_price, exit_price, stop_loss, take_profit,
                        open_time, close_time, trade_date, session,
                        profit_loss, net_profit, commission, swap,
                        notes, tags, status, source
                    ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, 'MT5')
                    ON CONFLICT (account_id, mt5_ticket) 
                    WHERE mt5_ticket IS NOT NULL
                    DO UPDATE SET
                        exit_price = COALESCE(EXCLUDED.exit_price, trades.exit_price),
                        close_time = COALESCE(EXCLUDED.close_time, trades.close_time),
                        profit_loss = COALESCE(EXCLUDED.profit_loss, trades.profit_loss),
                        net_profit = COALESCE(EXCLUDED.net_profit, trades.net_profit),
                        commission = COALESCE(EXCLUDED.commission, trades.commission),
                        swap = COALESCE(EXCLUDED.swap, trades.swap),
                        status = CASE WHEN EXCLUDED.exit_price IS NOT NULL THEN 'CLOSED' ELSE trades.status END,
                        updated_at = CURRENT_TIMESTAMP`,
                    [account.id, account.user_id, ticket, ticket, pair, direction, volume,
                        entryPrice, exitPrice, stopLoss, takeProfit, openTime, closeTime,
                        tradeDate, session, profitLoss || netProfit, netProfit,
                        commission, swap, notes, tags, exitPrice ? 'CLOSED' : 'OPEN']
                );
                syncedCount++;
            } catch (tradeError) {
                console.error('❌ Error syncing trade:', tradeError);
                errorCount++;
            }
        }

        console.log(`✅ Sync completed for account ${accountNumber}: ${syncedCount} synced, ${skippedCount} skipped, ${errorCount} errors`);

        res.json({
            message: 'Sync completed',
            syncedCount,
            skippedCount,
            errorCount,
            totalProcessed: trades.length
        });
    } catch (error) {
        console.error('❌ Sync error:', error);
        res.status(500).json({ error: 'Failed to sync trades' });
    }
});

// Delete all MT5 trades (for testing/reset)
router.delete('/clear-mt5', async (req, res) => {
    try {
        const result = await query("DELETE FROM trades WHERE source = 'MT5'");
        console.log(`🗑️ Deleted ${result.rowCount} MT5 trades`);
        res.json({ 
            message: 'MT5 trades cleared', 
            deletedCount: result.rowCount 
        });
    } catch (error) {
        console.error('❌ Clear error:', error);
        res.status(500).json({ error: 'Failed to clear trades' });
    }
});

module.exports = router;

File: backend_python/routes/trades.py
"""
Trading Journal - Trades Routes
แปลงจาก: backend/src/routes/trades.js
"""
from flask import Blueprint, request, session, jsonify
from functools import wraps
import sqlite3
import os
from datetime import datetime

trades_bp = Blueprint('journal_trades', __name__, url_prefix='/api/trading-journal/trades')

# Database path
DB_PATH = os.path.join(os.path.dirname(__file__), '../../../data/accounts.db')

def get_db():
    """Get database connection with Row factory"""
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

def login_required(f):
    """Decorator to require login"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        if 'user_id' not in session:
            return jsonify({'error': 'Unauthorized'}), 401
        return f(*args, **kwargs)
    return decorated_function

@trades_bp.route('', methods=['GET'])
@login_required
def get_trades():
    """
    GET /api/trading-journal/trades
    Get all trades for the authenticated user with filters
    """
    user_id = session.get('user_id')
    
    try:
        # Get query parameters
        status = request.args.get('status')
        pair = request.args.get('pair')
        account_id = request.args.get('account_id')
        start_date = request.args.get('startDate')
        end_date = request.args.get('endDate')
        limit = int(request.args.get('limit', 100))
        offset = int(request.args.get('offset', 0))
        
        conn = get_db()
        cursor = conn.cursor()
        
        # Build query
        query = 'SELECT * FROM journal_trades WHERE user_id = ?'
        params = [user_id]
        
        if account_id:
            query += ' AND account_id = ?'
            params.append(account_id)
        
        if status:
            query += ' AND status = ?'
            params.append(status)
        
        if pair:
            query += ' AND pair = ?'
            params.append(pair)
        
        if start_date:
            query += ' AND trade_date >= ?'
            params.append(start_date)
        
        if end_date:
            query += ' AND trade_date <= ?'
            params.append(end_date)
        
        query += ' ORDER BY open_time DESC LIMIT ? OFFSET ?'
        params.extend([limit, offset])
        
        cursor.execute(query, params)
        trades = [dict(row) for row in cursor.fetchall()]
        conn.close()
        
        return jsonify(trades)
        
    except Exception as e:
        print(f'❌ Get trades error: {e}')
        return jsonify({'error': 'Failed to fetch trades'}), 500

@trades_bp.route('/<int:trade_id>', methods=['GET'])
@login_required
def get_trade(trade_id):
    """
    GET /api/trading-journal/trades/<trade_id>
    Get specific trade
    """
    user_id = session.get('user_id')
    
    try:
        conn = get_db()
        cursor = conn.cursor()
        
        cursor.execute('''
            SELECT * FROM journal_trades 
            WHERE id = ? AND user_id = ?
        ''', (trade_id, user_id))
        
        trade = cursor.fetchone()
        conn.close()
        
        if not trade:
            return jsonify({'error': 'Trade not found'}), 404
        
        return jsonify(dict(trade))
        
    except Exception as e:
        print(f'❌ Get trade error: {e}')
        return jsonify({'error': 'Failed to fetch trade'}), 500

@trades_bp.route('', methods=['POST'])
def create_trade():
    """
    POST /api/trading-journal/trades
    Create new trade (manual or via MT5 EA webhook)
    Supports both JWT and accountNumber authentication
    """
    try:
        data = request.json
        account_number = data.get('accountNumber')
        
        # Authenticate via session or accountNumber
        user_id = None
        account_id = None
        
        if 'user_id' in session:
            # Authenticated via session
            user_id = session['user_id']
            account_id = data.get('account_id')  # Must be provided
            
            if not account_id:
                return jsonify({'error': 'account_id required'}), 400
                
        elif account_number:
            # Authenticate via accountNumber (EA webhook)
            conn = get_db()
            cursor = conn.cursor()
            
            cursor.execute('''
                SELECT id, user_id FROM mt5_accounts 
                WHERE account_number = ?
            ''', (account_number,))
            
            account = cursor.fetchone()
            conn.close()
            
            if not account:
                return jsonify({'error': 'MT5 account not found'}), 404
            
            user_id = account['user_id']
            account_id = account['id']
        else:
            return jsonify({'error': 'Authentication required'}), 401
        
        # Extract trade data
        event = data.get('event')
        mt5_ticket = data.get('mt5Ticket') or data.get('ticket')
        position_id = data.get('positionId') or data.get('mt5_position_id')
        pair = data.get('pair')
        direction = data.get('direction')
        volume = data.get('volume')
        entry_price = data.get('entryPrice') or data.get('entry_price')
        exit_price = data.get('exitPrice') or data.get('exit_price')
        stop_loss = data.get('stopLoss') or data.get('stop_loss')
        take_profit = data.get('takeProfit') or data.get('take_profit')
        profit_loss = data.get('profitLoss') or data.get('profit_loss')
        commission = data.get('commission', 0)
        swap = data.get('swap', 0)
        open_time = data.get('openTime') or data.get('open_time')
        close_time = data.get('closeTime') or data.get('close_time')
        trade_date = data.get('tradeDate') or data.get('trade_date')
        session_name = data.get('session')
        notes = data.get('notes', '')
        tags = data.get('tags', '')
        
        conn = get_db()
        cursor = conn.cursor()
        
        # Handle MT5 EA events
        if event == 'ORDER_OPEN':
            # Check if trade already exists
            if mt5_ticket:
                cursor.execute('''
                    SELECT id FROM journal_trades 
                    WHERE mt5_ticket = ? AND account_id = ?
                ''', (mt5_ticket, account_id))
                
                existing_trade = cursor.fetchone()
                
                if existing_trade:
                    # Update existing trade
                    cursor.execute('''
                        UPDATE journal_trades SET
                            pair = ?, direction = ?, volume = ?, entry_price = ?,
                            stop_loss = ?, take_profit = ?, open_time = ?, 
                            trade_date = ?, session = ?, status = 'OPEN', source = 'MT5',
                            updated_at = ?
                        WHERE id = ?
                    ''', (pair, direction, volume, entry_price, stop_loss, take_profit,
                          open_time, trade_date, session_name, datetime.now().isoformat(),
                          existing_trade['id']))
                    
                    conn.commit()
                    
                    cursor.execute('SELECT * FROM journal_trades WHERE id = ?', (existing_trade['id'],))
                    result = dict(cursor.fetchone())
                    conn.close()
                    
                    return jsonify(result), 200
            
            # Create new OPEN trade
            cursor.execute('''
                INSERT INTO journal_trades (
                    account_id, user_id, mt5_ticket, mt5_position_id, pair, direction, volume,
                    entry_price, stop_loss, take_profit, open_time, trade_date, session,
                    status, source
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'OPEN', 'MT5')
            ''', (account_id, user_id, mt5_ticket, position_id, pair, direction, volume,
                  entry_price, stop_loss, take_profit, open_time, trade_date, session_name))
            
            conn.commit()
            trade_id = cursor.lastrowid
            
            cursor.execute('SELECT * FROM journal_trades WHERE id = ?', (trade_id,))
            result = dict(cursor.fetchone())
            conn.close()
            
            return jsonify(result), 201
            
        elif event == 'ORDER_CLOSE':
            # Find and update OPEN trade to CLOSED
            if position_id:
                cursor.execute('''
                    SELECT id FROM journal_trades 
                    WHERE mt5_position_id = ? AND account_id = ? AND status = 'OPEN'
                ''', (position_id, account_id))
                
                existing_trade = cursor.fetchone()
                
                if existing_trade:
                    # Update to CLOSED
                    cursor.execute('''
                        UPDATE journal_trades SET
                            exit_price = ?, close_time = ?, profit_loss = ?,
                            commission = ?, swap = ?, status = 'CLOSED',
                            updated_at = ?
                        WHERE id = ?
                    ''', (exit_price, close_time, profit_loss, commission or 0, 
                          swap or 0, datetime.now().isoformat(), existing_trade['id']))
                    
                    conn.commit()
                    
                    cursor.execute('SELECT * FROM journal_trades WHERE id = ?', (existing_trade['id'],))
                    result = dict(cursor.fetchone())
                    conn.close()
                    
                    return jsonify(result), 200
                else:
                    # No matching OPEN trade found
                    conn.close()
                    print(f'⚠️ Skipping ORDER_CLOSE for position {position_id} - no matching OPEN trade')
                    return jsonify({
                        'message': 'CLOSE event skipped - no matching OPEN trade',
                        'positionId': position_id,
                        'note': 'Will be handled by historical sync'
                    }), 200
        
        # Manual trade creation
        status = 'CLOSED' if exit_price else 'OPEN'
        
        cursor.execute('''
            INSERT INTO journal_trades (
                account_id, user_id, pair, direction, volume, entry_price, exit_price,
                stop_loss, take_profit, open_time, close_time, trade_date, session,
                profit_loss, commission, swap, notes, tags, status, source
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'MANUAL')
        ''', (account_id, user_id, pair, direction, volume, entry_price, exit_price,
              stop_loss, take_profit, open_time, close_time, trade_date, session_name,
              profit_loss, commission, swap, notes, tags, status))
        
        conn.commit()
        trade_id = cursor.lastrowid
        
        cursor.execute('SELECT * FROM journal_trades WHERE id = ?', (trade_id,))
        result = dict(cursor.fetchone())
        conn.close()
        
        return jsonify(result), 201
        
    except Exception as e:
        print(f'❌ Create trade error: {e}')
        return jsonify({'error': 'Failed to create trade'}), 500

@trades_bp.route('/<int:trade_id>', methods=['PUT'])
@login_required
def update_trade(trade_id):
    """
    PUT /api/trading-journal/trades/<trade_id>
    Update existing trade
    """
    user_id = session.get('user_id')
    data = request.json
    
    try:
        conn = get_db()
        cursor = conn.cursor()
        
        # Build dynamic update query
        updates = []
        params = []
        
        fields = ['pair', 'direction', 'volume', 'entry_price', 'exit_price',
                  'stop_loss', 'take_profit', 'open_time', 'close_time', 'trade_date',
                  'session', 'profit_loss', 'commission', 'swap', 'notes', 'tags', 'status']
        
        for field in fields:
            if field in data:
                updates.append(f'{field} = ?')
                params.append(data[field])
        
        if not updates:
            conn.close()
            return jsonify({'error': 'No fields to update'}), 400
        
        # Add updated_at
        updates.append('updated_at = ?')
        params.append(datetime.now().isoformat())
        
        params.extend([trade_id, user_id])
        
        cursor.execute(f'''
            UPDATE journal_trades 
            SET {', '.join(updates)}
            WHERE id = ? AND user_id = ?
        ''', params)
        
        conn.commit()
        
        if cursor.rowcount == 0:
            conn.close()
            return jsonify({'error': 'Trade not found'}), 404
        
        # Fetch updated trade
        cursor.execute('SELECT * FROM journal_trades WHERE id = ?', (trade_id,))
        result = dict(cursor.fetchone())
        conn.close()
        
        return jsonify(result)
        
    except Exception as e:
        print(f'❌ Update trade error: {e}')
        return jsonify({'error': 'Failed to update trade'}), 500

@trades_bp.route('/<int:trade_id>', methods=['DELETE'])
@login_required
def delete_trade(trade_id):
    """
    DELETE /api/trading-journal/trades/<trade_id>
    Delete trade
    """
    user_id = session.get('user_id')
    
    try:
        conn = get_db()
        cursor = conn.cursor()
        
        cursor.execute('''
            DELETE FROM journal_trades 
            WHERE id = ? AND user_id = ?
        ''', (trade_id, user_id))
        
        conn.commit()
        deleted_count = cursor.rowcount
        conn.close()
        
        if deleted_count == 0:
            return jsonify({'error': 'Trade not found'}), 404
        
        return jsonify({'message': 'Trade deleted successfully'})
        
    except Exception as e:
        print(f'❌ Delete trade error: {e}')
        return jsonify({'error': 'Failed to delete trade'}), 500

@trades_bp.route('/bulk-delete', methods=['POST'])
@login_required
def bulk_delete_trades():
    """
    POST /api/trading-journal/trades/bulk-delete
    Bulk delete trades
    """
    user_id = session.get('user_id')
    data = request.json
    ids = data.get('ids', [])
    
    if not ids or not isinstance(ids, list) or len(ids) == 0:
        return jsonify({'error': 'Trade IDs required'}), 400
    
    try:
        conn = get_db()
        cursor = conn.cursor()
        
        # Create placeholders for IN clause
        placeholders = ','.join('?' * len(ids))
        
        cursor.execute(f'''
            DELETE FROM journal_trades 
            WHERE user_id = ? AND id IN ({placeholders})
        ''', [user_id] + ids)
        
        conn.commit()
        deleted_count = cursor.rowcount
        conn.close()
        
        return jsonify({
            'message': 'Trades deleted successfully',
            'deletedCount': deleted_count
        })
        
    except Exception as e:
        print(f'❌ Bulk delete error: {e}')
        return jsonify({'error': 'Failed to delete trades'}), 500

@trades_bp.route('/sync', methods=['POST'])
def sync_trades():
    """
    POST /api/trading-journal/trades/sync
    Sync multiple trades from MT5 (for initial sync or recovery)
    Supports accountNumber authentication
    """
    try:
        data = request.json
        account_number = data.get('accountNumber')
        trades = data.get('trades', [])
        
        if not account_number:
            return jsonify({'error': 'Account number required'}), 400
        
        if not trades or not isinstance(trades, list) or len(trades) == 0:
            return jsonify({'error': 'Trades array required'}), 400
        
        conn = get_db()
        cursor = conn.cursor()
        
        # Find account by accountNumber
        cursor.execute('''
            SELECT id, user_id FROM mt5_accounts 
            WHERE account_number = ?
        ''', (account_number,))
        
        account = cursor.fetchone()
        
        if not account:
            conn.close()
            return jsonify({'error': 'Account not found'}), 404
        
        account_id = account['id']
        user_id = account['user_id']
        
        synced_count = 0
        error_count = 0
        
        for trade in trades:
            try:
                ticket = trade.get('ticket')
                
                if not ticket:
                    error_count += 1
                    continue
                
                # Check if trade already exists
                cursor.execute('''
                    SELECT id FROM journal_trades 
                    WHERE mt5_ticket = ? AND account_id = ?
                ''', (ticket, account_id))
                
                existing = cursor.fetchone()
                
                exit_price = trade.get('exitPrice')
                status = 'CLOSED' if exit_price else 'OPEN'
                
                if existing:
                    # Update existing trade
                    cursor.execute('''
                        UPDATE journal_trades SET
                            exit_price = COALESCE(?, exit_price),
                            close_time = COALESCE(?, close_time),
                            profit_loss = COALESCE(?, profit_loss),
                            commission = COALESCE(?, commission),
                            swap = COALESCE(?, swap),
                            status = ?,
                            updated_at = ?
                        WHERE id = ?
                    ''', (exit_price, trade.get('closeTime'), trade.get('profitLoss'),
                          trade.get('commission'), trade.get('swap'), status,
                          datetime.now().isoformat(), existing['id']))
                else:
                    # Insert new trade
                    cursor.execute('''
                        INSERT INTO journal_trades (
                            account_id, user_id, mt5_ticket, mt5_position_id, pair, direction, volume,
                            entry_price, exit_price, stop_loss, take_profit,
                            open_time, close_time, trade_date, session,
                            profit_loss, commission, swap,
                            notes, tags, status, source
                        ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'MT5')
                    ''', (account_id, user_id, ticket, ticket, trade.get('pair'),
                          trade.get('direction'), trade.get('volume'), trade.get('entryPrice'),
                          exit_price, trade.get('stopLoss'), trade.get('takeProfit'),
                          trade.get('openTime'), trade.get('closeTime'), trade.get('tradeDate'),
                          trade.get('session'), trade.get('profitLoss'), trade.get('commission'),
                          trade.get('swap'), trade.get('notes', ''), trade.get('tags', ''), status))
                
                synced_count += 1
                
            except Exception as trade_error:
                print(f'❌ Error syncing trade: {trade_error}')
                error_count += 1
        
        conn.commit()
        conn.close()
        
        print(f'✅ Sync completed for account {account_number}: {synced_count} synced, {error_count} errors')
        
        return jsonify({
            'message': 'Sync completed',
            'syncedCount': synced_count,
            'errorCount': error_count,
            'totalProcessed': len(trades)
        })
        
    except Exception as e:
        print(f'❌ Sync error: {e}')
        return jsonify({'error': 'Failed to sync trades'}), 500

@trades_bp.route('/clear-mt5', methods=['DELETE'])
def clear_mt5_trades():
    """
    DELETE /api/trading-journal/trades/clear-mt5
    Delete all MT5 trades (for testing/reset)
    """
    try:
        conn = get_db()
        cursor = conn.cursor()
        
        cursor.execute("DELETE FROM journal_trades WHERE source = 'MT5'")
        
        conn.commit()
        deleted_count = cursor.rowcount
        conn.close()
        
        print(f'🗑️ Deleted {deleted_count} MT5 trades')
        
        return jsonify({
            'message': 'MT5 trades cleared',
            'deletedCount': deleted_count
        })
        
    except Exception as e:
        print(f'❌ Clear error: {e}')
        return jsonify({'error': 'Failed to clear trades'}), 500
